File: java/leetcode_grind/Day785.java
package leetcode_grind;

import java.util.Stack;

public class Day785 {
    // https://leetcode.com/problems/check-if-a-parentheses-string-can-be-valid/description/?envType=daily-question&envId=2025-01-12
    static class Solution1 {

        public boolean canBeValid(String s, String locked) {
            int length = s.length();

            // If length of string is odd, return false.
            if (length % 2 == 1) {
                return false;
            }

            Stack<Integer> openBrackets = new Stack<>();
            Stack<Integer> unlocked = new Stack<>();

            // Iterate through the string to handle '(' and ')'
            for (int i = 0; i < length; i++) {
                if (locked.charAt(i) == '0') {
                    unlocked.push(i);
                } else if (s.charAt(i) == '(') {
                    openBrackets.push(i);
                } else if (s.charAt(i) == ')') {
                    if (!openBrackets.empty()) {
                        openBrackets.pop();
                    } else if (!unlocked.empty()) {
                        unlocked.pop();
                    } else {
                        return false;
                    }
                }
            }

            // Match remaining open brackets with unlocked characters
            while (!openBrackets.empty() &&
                    !unlocked.empty() &&
                    openBrackets.peek() < unlocked.peek()) {
                openBrackets.pop();
                unlocked.pop();
            }

            if (!openBrackets.empty()) {
                return false;
            }

            return true;
        }
    }

}

File: src/leetcode_grind/day_785.rs
// https://leetcode.com/problems/check-if-a-parentheses-string-can-be-valid/description/?envType=daily-question&envId=2025-01-12
pub fn can_be_valid(s: String, locked: String) -> bool {
    let s = s.chars().collect::<Vec<_>>();
    let locked = locked.chars().collect::<Vec<_>>();

    if s.len() % 2 == 1 {
        return false;
    }

    let mut open = 0;
    let mut unlocked = 0;

    for i in 0..s.len() {
        if locked[i] == '0' {
            unlocked += 1;
        } else if s[i] == '(' {
            open += 1;
        } else if s[i] == ')' {
            if open > 0 {
                open -= 1;
            } else if unlocked > 0 {
                unlocked -= 1;
            } else {
                return false;
            }
        }
    }

    let mut balance = 0;
    for i in (0..s.len()).rev() {
        if locked[i] == '0' {
            balance -= 1;
            unlocked -= 1;
        } else if s[i] == '(' {
            balance += 1;
            open -= 1;
        } else if s[i] == ')' {
            balance -= 1;
        }

        if balance > 0 {
            return false;
        }

        if unlocked == 0 && open == 0 {
            break;
        }
    }

    if open > 0 {
        return false;
    }
    true
}
